### ISSUE TITLE
Confusing error: access of union field 'a' while field 'a' is active

### ISSUE THREAD

**Issue by ssmid** [ziglang/zig#19211]
### Zig Version

0.12.0-dev.1871+e426ae43a

### Steps to Reproduce and Observed Output

Code:
```zig
const std = @import("std");

const U = union {
        a: u32,
        b: u32,
};

pub fn main() void {
        var u: U = undefined;
        u.a = 0;
        std.debug.print("test: {d}\n", .{u.a});
}
```
Output:
`thread 49835 panic: access of union field 'a' while field 'a' is active`

### Expected Output

`thread 49835 panic: assignment to union field 'a' while no field is active`

Edit: removed last two lines which were unnecessary

**Comment by RossComputerGuy** [comment]
Does the error go away if you don't set the variables to undefined?

**Comment by RossComputerGuy** [comment]
Yeah, to change the union tag which is active, you have to reinitialize it. This isn't really bad behavior, the initial error message of using undefined and then setting the value could be improved. However, the underlying behavior will likely not be changed. Using undefined this way will always result in some sort of error.

**Comment by nektro** [comment]
I think this is a bug instead of a bad error message. if there is no tag set, then assigning it should work imo.

**Comment by RossComputerGuy** [comment]
@nektro Eh, I don't think so because then the field builtin should work for initialization but we always use the union init builtin. If this were to not be an error, would that mean the field builtin would work and the union init builtin should be removed?

**Comment by nektro** [comment]
yes if `.` were to work then `@field` would too. its either a bug or an illegal use of `undefined` like you hinted at above; im torn on which one though.

**Comment by xdBronch** [comment]
the safety check is a branch, effectively `if (accessed != current) @panic()` so i'd definitely say this is illegal behavior. the error message could certainly be improved

**Comment by RossComputerGuy** [comment]
@nektro I'd say keep the existing behavior as we already have a solution and accessing an inactive tag is considered UB.

@xdBronch I agree, it should be something like "error: unexpected undefined, wants union".

**Comment by nektro** [comment]
detecting undefined is not possible

**Comment by RossComputerGuy** [comment]
@nektro True but from the panic error, it says a is active but a is being accessed but it's not active. It should be possible to detect this weird state and report that you're using undefined instead of union init or something else.

**Comment by xdBronch** [comment]
the safety check is in ReleaseSafe as well yes

**Comment by nektro** [comment]
new twist: on zig `0.12.0-dev.1828+225fe6ddb` it prints `test: 0`

**Comment by dravenk** [comment]
In version 0.14.0, there is no panic return.
```
dk@dkmac /tmp % zig run t1.zig
test: 0
dk@dkmac /tmp % cat t1.zig 
const std = @import("std");

const U = union {
        a: u32,
        b: u32,
};

pub fn main() void {
        var u: U = undefined;
        u.a = 0;
        std.debug.print("test: {d}\n", .{u.a});
}
dk@dkmac /tmp % zig version
0.14.0-dev.2198+e5f5229fd
```
<img width="352" alt="image" src="https://github.com/user-attachments/assets/828f2332-7a3c-4faf-ba01-5737db316962">

**Comment by xdBronch** [comment]
this is undefined behavior, there's no guarantee that it will or won't work or how

**Comment by hvbtup** [comment]
I get this error message also in this code:
https://codeberg.org/tumirnix/exla/src/commit/0e5851e9da91b73cb3a347899f7030d89aa3cbee/moin/src/object.zig#L1248

In my case, the union `items` was initialized in line 1261 before being accessed, so I don't understand how this could be UB.
Anyway, the error message is very confusing.

**Comment by nektro** [comment]
bare unions are secretly tagged in debug mode. the correct initialization is:
`tuple.items = .{ .direct = .{ Value.NIL, Value.NIL } };`
